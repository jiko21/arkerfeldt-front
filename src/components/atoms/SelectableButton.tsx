/** @jsxImportSource @emotion/react */
import { css } from '@emotion/react';
import { FC, useRef, useState } from 'react';
import { Color } from '@/const/color';
import styled from '@emotion/styled';
import { faChevronUp, faChevronDown } from '@fortawesome/free-solid-svg-icons';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';

const buttonStyle = css`
  align-items: center;
  background-color: ${Color.ENABLE};
  border: none;
  box-shadow: 4px 4px 4px rgba(0, 0, 0, 0.25);
  color: #ffffff;
  display: flex;
  font-size: 14px;
  font-weight: 700;
  height: 30px;
  justify-content: flex-start;
  min-width: 240px;
  padding: 0;

  &:hover {
    background-color: ${Color.ENABLE_HOVER};
  }
`;

const svgWrapper = css`
  align-items: center;
  display: flex;
  border-right: 1px solid ${Color.WHITE};
  height: 100%;
  justify-content: center;
  width: 30px;
`;

const childWrapper = css`
  align-items: center;
  display: flex;
  justify-content: center;
  width: 100%;
`;

const Ul = styled.ul<{ width?: number }>`
  border-radius: 6px;
  filter: drop-shadow(0px 4px 4px rgba(0, 0, 0, 0.25));
  list-style: none;
  margin: 0;
  padding: 0;
  position: fixed;
  width: ${(props) => (props.width ? props.width : 240)}px;
`;

const listStyle = css`
  background: ${Color.WHITE_LIGHT};
  cursor: pointer;
  padding: 4px;

  :hover {
    background: ${Color.WHITE_LIGHT_HOVER};
  }
`;

type Candidate = {
  text: string;
  value: string;
};

type Props = {
  onClick: () => void;
  onItemClick: (candidate: Candidate) => void;
  candidates: Candidate[];
};

const SelectableButton: FC<Props> = ({
  onClick,
  onItemClick,
  children,
  candidates,
}) => {
  const [isSelect, changeSelectStatus] = useState(false);
  const buttonRef = useRef<HTMLButtonElement>(null);
  return (
    <div>
      <button ref={buttonRef} css={buttonStyle} data-testid="enable-btn">
        <span
          css={svgWrapper}
          data-testid="button-tapped"
          onClick={() => {
            changeSelectStatus(!isSelect);
          }}
        >
          <FontAwesomeIcon icon={isSelect ? faChevronUp : faChevronDown} />
        </span>
        <div css={childWrapper} onClick={onClick} data-testid="button-area">
          {children}
        </div>
      </button>
      {isSelect && (
        <Ul
          width={buttonRef.current ? buttonRef.current.offsetWidth : undefined}
          data-testid="selectable-area"
        >
          {candidates.map((item) => (
            <li
              css={listStyle}
              onClick={() => {
                onItemClick(item);
                changeSelectStatus(!isSelect);
              }}
              key={item.value}
              data-testid={`candidate-${item.value}`}
            >
              {item.text}
            </li>
          ))}
        </Ul>
      )}
    </div>
  );
};

export default SelectableButton;
