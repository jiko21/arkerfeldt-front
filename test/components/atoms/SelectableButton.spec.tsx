import { fireEvent, render, screen } from '@testing-library/react';
import SelectableButton from '@/components/atoms/SelectableButton';

describe('SelectableButton.tsx', () => {
  const onClickMock = jest.fn();
  const onItemClickMock = jest.fn();

  const CANDIDATES = [
    {
      text: '未公開のまま保存',
      value: 'UNPUBLISHED',
    },
    {
      text: '公開する',
      value: 'PUBLISH',
    },
  ];

  it('should correctly render when list is closed', () => {
    const { container } = render(
      <SelectableButton
        onClick={onClickMock}
        onItemClick={onItemClickMock}
        candidates={CANDIDATES}
      >
        Sample
      </SelectableButton>,
    );
    expect(screen.getByText('Sample')).toBeTruthy();
    expect(screen.queryByTestId('selectable-area')).toBeFalsy();
    expect(container).toMatchSnapshot();
  });

  it('should correctly render when list is open', () => {
    const { container } = render(
      <SelectableButton
        onClick={onClickMock}
        onItemClick={onItemClickMock}
        candidates={CANDIDATES}
      >
        Sample
      </SelectableButton>,
    );

    fireEvent.click(screen.getByTestId('button-tapped'));
    expect(screen.getByText('Sample')).toBeTruthy();
    expect(screen.queryByTestId('selectable-area')).toBeTruthy();
    expect(container).toMatchSnapshot();

    CANDIDATES.forEach((item) => {
      expect(screen.getByText(item.text)).toBeTruthy();
    });
  });

  it('should correctly call `onItemClick`', () => {
    render(
      <SelectableButton
        onClick={onClickMock}
        onItemClick={onItemClickMock}
        candidates={CANDIDATES}
      >
        Sample
      </SelectableButton>,
    );

    fireEvent.click(screen.getByTestId('button-tapped'));
    fireEvent.click(screen.getByTestId('candidate-PUBLISH'));

    expect(onItemClickMock).toBeCalled();
  });

  it('should correctly call `onClick`', () => {
    render(
      <SelectableButton
        onClick={onClickMock}
        onItemClick={onItemClickMock}
        candidates={CANDIDATES}
      >
        Sample
      </SelectableButton>,
    );

    fireEvent.click(screen.getByTestId('button-area'));

    expect(onClickMock).toBeCalled();
  });
});
