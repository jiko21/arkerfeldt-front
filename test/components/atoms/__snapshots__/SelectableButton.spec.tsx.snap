// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SelectableButton.tsx should correctly render when list is closed 1`] = `
.emotion-0 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background-color: #007AAE;
  border: none;
  box-shadow: 4px 4px 4px rgba(0, 0, 0, 0.25);
  color: #ffffff;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  font-size: 14px;
  font-weight: 700;
  height: 30px;
  -webkit-box-pack: start;
  -ms-flex-pack: start;
  -webkit-justify-content: flex-start;
  justify-content: flex-start;
  min-width: 240px;
  padding: 0;
}

.emotion-0:hover {
  background-color: #005b82;
}

.emotion-1 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  border-right: 1px solid #FFFFFF;
  height: 100%;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  width: 30px;
}

.emotion-2 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  width: 100%;
}

<div>
  <div>
    <button
      class="emotion-0"
      data-testid="enable-btn"
    >
      <span
        class="emotion-1"
        data-testid="button-tapped"
      >
        <svg
          aria-hidden="true"
          class="svg-inline--fa fa-chevron-down fa-w-14 "
          data-icon="chevron-down"
          data-prefix="fas"
          focusable="false"
          role="img"
          viewBox="0 0 448 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M207.029 381.476L12.686 187.132c-9.373-9.373-9.373-24.569 0-33.941l22.667-22.667c9.357-9.357 24.522-9.375 33.901-.04L224 284.505l154.745-154.021c9.379-9.335 24.544-9.317 33.901.04l22.667 22.667c9.373 9.373 9.373 24.569 0 33.941L240.971 381.476c-9.373 9.372-24.569 9.372-33.942 0z"
            fill="currentColor"
          />
        </svg>
      </span>
      <div
        class="emotion-2"
        data-testid="button-area"
      >
        Sample
      </div>
    </button>
  </div>
</div>
`;

exports[`SelectableButton.tsx should correctly render when list is open 1`] = `
.emotion-0 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background-color: #007AAE;
  border: none;
  box-shadow: 4px 4px 4px rgba(0, 0, 0, 0.25);
  color: #ffffff;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  font-size: 14px;
  font-weight: 700;
  height: 30px;
  -webkit-box-pack: start;
  -ms-flex-pack: start;
  -webkit-justify-content: flex-start;
  justify-content: flex-start;
  min-width: 240px;
  padding: 0;
}

.emotion-0:hover {
  background-color: #005b82;
}

.emotion-1 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  border-right: 1px solid #FFFFFF;
  height: 100%;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  width: 30px;
}

.emotion-2 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  width: 100%;
}

.emotion-3 {
  border-radius: 6px;
  -webkit-filter: drop-shadow(0px 4px 4px rgba(0, 0, 0, 0.25));
  filter: drop-shadow(0px 4px 4px rgba(0, 0, 0, 0.25));
  list-style: none;
  margin: 0;
  padding: 0;
  position: fixed;
  width: 240px;
}

.emotion-4 {
  background: #FCFCFC;
  cursor: pointer;
  padding: 4px;
}

.emotion-4:hover {
  background: #ECECEC;
}

<div>
  <div>
    <button
      class="emotion-0"
      data-testid="enable-btn"
    >
      <span
        class="emotion-1"
        data-testid="button-tapped"
      >
        <svg
          aria-hidden="true"
          class="svg-inline--fa fa-chevron-up fa-w-14 "
          data-icon="chevron-up"
          data-prefix="fas"
          focusable="false"
          role="img"
          viewBox="0 0 448 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M240.971 130.524l194.343 194.343c9.373 9.373 9.373 24.569 0 33.941l-22.667 22.667c-9.357 9.357-24.522 9.375-33.901.04L224 227.495 69.255 381.516c-9.379 9.335-24.544 9.317-33.901-.04l-22.667-22.667c-9.373-9.373-9.373-24.569 0-33.941L207.03 130.525c9.372-9.373 24.568-9.373 33.941-.001z"
            fill="currentColor"
          />
        </svg>
      </span>
      <div
        class="emotion-2"
        data-testid="button-area"
      >
        Sample
      </div>
    </button>
    <ul
      class="emotion-3"
      data-testid="selectable-area"
      width="0"
    >
      <li
        class="emotion-4"
        data-testid="candidate-UNPUBLISHED"
      >
        未公開のまま保存
      </li>
      <li
        class="emotion-4"
        data-testid="candidate-PUBLISH"
      >
        公開する
      </li>
    </ul>
  </div>
</div>
`;
